libraries/chain_kv/include/b1/session/session.hpp

```diff
@@ -0,0 +1,1061 @@
+    {
+        session_impl();
+        session_impl(session_impl& parent_);
+        session_impl(persistent_data_store pds);
+        session_impl(persistent_data_store pds, cache_data_store cds);
+        ~session_impl();
+        
+        auto undo() -> void;
+        auto commit() -> void;
+ 
+        std::shared_ptr<session_impl> parent;
+        std::weak_ptr<session_impl> child;
+        
+        // The persistent data store.  This is shared across all levels of the session.
+        std::shared_ptr<persistent_data_store> backing_data_store;
+        
+        // The cache used by this session instance.  This will include all new/update key_values
+        // and may include values read from the persistent data store.
+        cache_data_store cache;
+        // keys that have been updated during this session.
+        std::unordered_set<bytes> updated_keys;
+        // keys that have been deleted during this session.
+        std::unordered_set<bytes> deleted_keys;
+    };
+    
+    // session implement the PIMPL idiom to allow for stack based semantics.
+    std::shared_ptr<session_impl> m_impl;
+};
+
+template <typename persistent_data_store, typename cache_data_store>
+auto make_session() -> session<persistent_data_store, cache_data_store>
+{
+    using session_type = session<persistent_data_store, cache_data_store>;
+    return session_type{};
+}
+
+template <typename persistent_data_store, typename cache_data_store>
+auto make_session(persistent_data_store store) -> session<persistent_data_store, cache_data_store>
+{
+    using session_type = session<persistent_data_store, cache_data_store>;
+    return session_type{std::move(store)};
+}
+
+template <typename persistent_data_store, typename cache_data_store>
+auto make_session(persistent_data_store store, cache_data_store cache) -> session<persistent_data_store, cache_data_store>
+{
+    using session_type = session<persistent_data_store, cache_data_store>;
+    return session_type{std::move(store), std::move(cache)};
+}
+
+template <typename persistent_data_store, typename cache_data_store>
+auto make_session(session<persistent_data_store, cache_data_store>& the_session) -> session<persistent_data_store, cache_data_store>
+{
+    using session_type = session<persistent_data_store, cache_data_store>;
+    return session_type{the_session, typename session_type::nested_session_t{}};
+}
+
+template <typename persistent_data_store, typename cache_data_store>
+session<persistent_data_store, cache_data_store>::session_impl::session_impl()
+: backing_data_store{std::make_shared<persistent_data_store>()}
+{
+}
+
+template <typename persistent_data_store, typename cache_data_store>
+session<persistent_data_store, cache_data_store>::session_impl::session_impl(session_impl& parent_)
+: parent{parent_.shared_from_this()},
+  backing_data_store{parent_.backing_data_store}
+{
+    if (auto child_ptr = parent->child.lock())
+    {
+        child_ptr->backing_data_store = nullptr;
```

Comment:
Should comment how the current logic means that the current child will be abandoned (undone).  The other option is to throw an exception if the parent session already has a child.